### DISCUSSION TITLE
Workaround for "not enough context to apply patch"

### DISCUSSION THREAD

**Discussion by dabrahams** [magit/magit#4952] [Support] [answered]
Sometimes I get myself into a situation where I've made a lot of local changes and need to sort them into separate commits, and I frequently can't make the diff context small enough so that I can stage just the parts I want.  I don't really know what the feature would look like, but it would be awesome to have a workaround for these situations.  Maybe a way to directly edit the notional copy-of-the-file-in-the-index?

**Comment by tarsius** [comment]
Often it is possible to mark the part that you want to stage using the region before staging, but I assume you know that. In really hairy cases, I sometimes end up reverting some changes by manually editing the file (or using undo, if I get lucky), staging, and then undoing the edits/undos.

**Reply by dabrahams** [reply]
> Often it is possible to mark the part that you want to stage using the region before staging, but I assume you know that

Actually I don't think I understand what you mean.

**Reply by tarsius** [reply, marked as answer]
I mean this (from https://magit.vc/manual/magit/Staging-and-Unstaging.html):

> With Magit, on the other hand, one can un-/stage individual hunks by
> just moving point into the respective section inside a diff displayed in
> the status buffer or a separate diff buffer and typing ‘s’ or ‘u’.  To
> operate on just parts of a hunk, mark the changes that should be
> un-/staged using the region and then press the same key that would be
> used to un-/stage.  To stage multiple files or hunks at once use a
> region that starts inside the heading of such a section and ends inside
> the heading of a sibling section of the same type.

Note that while you can stage multiple hunks at once, by marking a region that begins in the header of one hunk and ends in the hunk of another (of the same file), you cannot stage the second half of one hunk and the first half of the next hunk together. The region must begin and end in the same hunk in this case. Also note how the region is visualized differently depending on what kind of selection you make. If it looks like in any other place, then it is not a valid selection as far as these commands are concerned.

**Comment by dabrahams** [comment]
Oh, this is great to know about! No, I wasn't aware of it!

**Reply by dabrahams** [reply]
… And I forgot about this completely before I got a chance to use it. Great that it showed up in my inbox to remind me. I hope I'll remember it when I do need it which leads me to suggest: Maybe something should be done to make things like this more discoverable (tool tips?) I realize this is a very general suggestion and probably not actionable, but I thought maybe I should raise it.
